### PR TITLE
Goals Screen: Save onboarding goals when creating site

### LINKED ISSUE #96755: Persist Goal selection from pre-account Goals screen
Goals will be pre-account creation and thus must be persisted through the flow and then saved when the user has an account.

## Prior work
This is for information only, to review what has been done before
 
**Trail Map**
Work in Trail Map project is relevant to moving the goals screen: pdDR7T-1xi-p2
From @wongasy (We won't have a survey but the mechanics could be the same for recording goals.)

We can define survey questions and answers: uggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Suggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Sfbhepr%2Skers%2Sjcpbz%2Sjc%2Qpbagrag%2Syvo%2Sfrtzragngvba%2Qfheirlf%2S-og
And use this component for rendering the survey: uggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Suggcf%3N%2S%2Suers.yv%2S%3Suggcf%3N%2S%2Sbcratebx.n8p.pbz%2Sfbhepr%2Skers%2Spnylcfb%2Spyvrag%2Spbzcbaragf%2Sfrtzragngvba%2Qfheirl%2Svaqrk.gfk%3Se%3Q50o9pnsr%2361-og (or create another component if the UI looks different, but the hook should still be reusable)

**Goals stored in site option**
And Phil's investigation here: pfYzsZ-w2-p2

### PR DESCRIPTION
**Pull request by vykes-mac**
Closes https://github.com/Automattic/wp-calypso/issues/96755

## Proposed Changes
This PR allows the selected goals in the onboarding flow to be persisted in the sites creation step. The `/sites/new` endpoint accepts an option `site_goals` which is used to flagged goals on the site, Here we set that option to the selected goals.

* Include `site_goals` option in the site creation payload for the onboarding flow.

## Testing Instructions

* go to `/setup/onboarding`
* Select list of goals and complete the flow (You can use a free plan)
* After site is created go to the blog RC and verify the goals have been added to the site
* You should see something similar to the image  
image

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

### CODE REVIEW

File: packages/onboarding/src/cart/index.tsx [hunk 1]
```diff
@@ -181,6 +184,8 @@ export const createSiteWithCart = async (
 	// 	return;
 	// }
 
+	const isOnboarding = isOnboardingFlow( flowName );
```
Comment (vykes-mac) [inline packages/onboarding/src/cart/index.tsx]: Maybe we could remove this check and only check if site goals available?

Comment (p-jackson) [reply packages/onboarding/src/cart/index.tsx]: I'm inclined to keep this check in place. Maybe even reinforce it with a check of `isGoalsFirstExperiment`

[There are issues like this](https://github.com/Automattic/wp-calypso/issues/96481#issuecomment-2487492997) where sometimes there are phantom goals left in the store from previous bailed user attempts. I wouldn't want to set those old phantom goals that just happen to be in the store onto the site.

Comment (p-jackson) [reply packages/onboarding/src/cart/index.tsx]: In order to get access to the `isGoalsFirstExperiment` value, maybe it'd be easier to always pass the `site_goals`  if they're passed to `createSiteWithCart`. And then to do the `isOnboardingFlow && isGoalsFirstExperiment` up in the `create-site` step? You've got more easy access to selectors and hooks in the step component.

**Review by p-jackson — CHANGES_REQUESTED** [review]
I made some comments below about how maybe we should ensure goals are only saved in this specific experiment. However I also discovered a bug if you go through the whole flow.

If you check blog RC immediately after site creation then you see the goals as you described. However I then completed the entire flow (i.e. completing `site-setup` and landing in the launchpad). Then after refreshing blog RC the goals have been cleared.

CleanShot 2024-12-13 at 10 38 40@2x

I know there's code in `site-setup` flow for setting goals. I would have thought it would just re-save the same goals. But it seems something in that flow is actually clearing out the site goals.

**Comment by vykes-mac** [comment]
> If you check blog RC immediately after site creation then you see the goals as you described. However I then completed the entire flow (i.e. completing site-setup and landing in the launchpad). Then after refreshing blog RC the goals have been cleared.

It seems the site setup exit flow is running the onboarding customisation endpoint twice resulting in the goals being cleared and so the second execution empties the goals.

https://github.com/Automattic/wp-calypso/blob/6c47bc28804bac8d716b98cd35494771db4c164d/client/landing/stepper/declarative-flow/site-setup-flow.ts#L251

File: client/landing/stepper/declarative-flow/site-setup-flow.ts [hunk 2]
```diff
@@ -214,7 +214,7 @@ const siteSetupFlow: Flow = {
 
 					const settings = {
 						site_intent: siteIntent,
-						site_goals: goals,
+						...( goals.length && { site_goals: goals } ),
```
Comment (vykes-mac) [inline client/landing/stepper/declarative-flow/site-setup-flow.ts:217]: @p-jackson The exit flow here got executed multiple times and so the `resetOnboardStoreWithSkipFlags` is called which cleared the goals. When its called again the goals are already clear and so it run onboarding customisation endpoint with empty goals. 

This is a quick fix to only include goals in the customisation if goals are present.

Comment (p-jackson) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:217]: 🤔 I think this is ok ...
Hypothetically it means that the `site-setup` flow can no longer clear the goals of an existing site. But the flow isn't really used that way (I jump into the flow with an existing site using testing, but it's not really a user flow). So this is probably fine.

Comment (p-jackson) [reply client/landing/stepper/declarative-flow/site-setup-flow.ts:217]: When the user clicks this button on the fullscreen launchpad they're taken back to the design picker

CleanShot 2024-12-13 at 12 33 41@2x

I tested to make sure this PR doesn't change this behaviour. It doesn't. I thought it took the user back to the `site-setup` flow, but it's actually it's own `update-design` flow.

**Review by p-jackson — APPROVED** [review]
lgtm

**Pull request merged**